go/tools/builders/vet.go

```diff
@@ -0,0 +1,184 @@
+	for _, m := range errMsgs {
+		if !vetErrorMsgPattern.MatchString(m) {
+			vetErrors = append(vetErrors, m)
+			continue
+		}
+		vetFindings = append(vetFindings, m)
+	}
+	if len(vetFindings) == 0 && len(vetErrors) != 0 {
+		return "", errors.New(strings.Join(vetErrors, "\n"))
+	}
+	return strings.Join(vetFindings, "\n"), nil
+}
+
+var vetFlags = []string{
+	// NOTE: Keep in sync with github.com/golang/go/src/cmd/go/internal/test/test.go
+	"-atomic",
+	"-bool",
+	"-buildtags",
+	"-nilfunc",
+	"-printf",
+}
+
+// vetErrorMsgPattern matches an error message emitted by vet.
+//
+// This regexp should be strict enough to exclude internal errors (prefixed with
+// "vet") and type-check errors (which additionally include a column number).
+//
+// NOTE: this might break if the formating of vet error messages changes.
+var vetErrorMsgPattern = regexp.MustCompile(`^([^:]+?):([0-9]+): (.*)`)
+
+// splitOutput was adapted from go/test/run.go.
+func splitOutput(out string) []string {
+	// gc error messages continue onto additional lines with leading tabs.
+	// Split the output at the beginning of each line that doesn't begin with a tab.
+	// <autogenerated> lines are impossible to match so those are filtered out.
+	var res []string
+	for _, line := range strings.Split(out, "\n") {
+		line = strings.TrimSuffix(line, "\r") // normalize Windows output
+		if strings.HasPrefix(line, "\t") {
+			res[len(res)-1] += "\n" + line
+		} else if strings.HasPrefix(line, "go tool") || strings.HasPrefix(line, "#") {
+			continue
+		} else if strings.TrimSpace(line) != "" {
+			res = append(res, line)
+		}
+	}
+	return res
+}
+
+// parseArchivesAndImportMap parses the -archivefile and -importmap arguments
+// created by the compile rule into the packageFiles and importMap maps used
+// by the vet and checker importers.
+func parseArchivesAndImportMap(archiveFiles, importMapIn []string) (packageFiles map[string]string, importMap map[string]string, err error) {
+	packageFiles, importMap = make(map[string]string), make(map[string]string)
+	for _, mapping := range importMapIn {
+		i := strings.Index(mapping, "=")
+		if i < 0 {
+			err = fmt.Errorf("invalid importmap %v: no = separator", mapping)
+			return
+		}
+		source := mapping[:i]
+		actual := mapping[i+1:]
+		if source == "" || actual == "" {
+			continue
+		}
+		importMap[source] = actual
+	}
+	for _, a := range archiveFiles {
+		kv := strings.Split(a, "=")
+		if len(kv) != 2 {
+			err = fmt.Errorf("invalid archivefile %v: no = separator", a)
+			return
+		}
+		path := kv[0]
+		if p, ok := importMap[path]; ok {
+			path = p
+		}
+		packageFiles[path] = kv[1]
+	}
+	return
+}
+
+// makeVetCfg creates a vet.cfg file and returns its file path. It is the
+// caller's responsibility to remove this file when it is no longer needed.
+func makeVetCfg(packageList, stdLib string, archiveFiles, importMap, files []string) (string, error) {
+	pf, im, err := parseArchivesAndImportMap(archiveFiles, importMap)
```

Comment:
`parseArchivesAndImportMap` does a fair amount of work, but it's called twice when vet is enabled. Can its output be passed into `makeVetCfg` instead?

Reply:
Done.